File: matlab/examples/cqrrpt_demo.m
% cqrrpt_demo  Rank revelation with RandLAPACK CQRRPT from MATLAB.
%
% CQRRPT discovers the numerical rank of a tall matrix rather than taking
% it as a parameter. This demo builds a matrix of known rank by repeating
% columns, then checks that the detected rank, the pivots, and the residual
% all agree with the construction.
%
% The shape below is chosen where the randomized method actually pays: dense
% QRCP wins on small problems (measured 0.6x at 5000-by-50) and loses as the
% matrix grows (7.4x at 20000-by-200 on the machine this was written on).
%
% Run from MATLAB after:
%   addpath('/path/to/randlapack-bindings/matlab')

m = 20000; r = 160; extra = 40;
B = randn(m, r);
A = [B, B(:, 1:extra)];                 % rank r, r+extra columns
perm = randperm(size(A, 2));
A = A(:, perm);                         % scatter the duplicates

t = tic;
[Q, R, J] = randlapack.cqrrpt(A);
dt = toc(t);

k = size(R, 1);
fprintf('A is %d-by-%d, constructed rank %d\n', m, size(A, 2), r);
fprintf('CQRRPT detected rank %d in %.3f s\n', k, dt);
fprintf('||A(:, J) - Q*R||_F / ||A||_F = %.2e\n', ...
        norm(A(:, J) - Q * R, 'fro') / norm(A, 'fro'));
fprintf('||Q^T Q - I||_F               = %.2e\n', ...
        norm(Q' * Q - eye(k), 'fro'));

% The first k pivots should select an independent set. Checking against
% MATLAB's own rank() is the honest comparison.
fprintf('rank(A(:, J(1:k)))            = %d  (want %d)\n', ...
        rank(A(:, J(1:k))), k);

% Compare against MATLAB's dense QRCP on the same matrix.
t = tic; [~, ~, Jd] = qr(A, 'vector'); dt_dense = toc(t);
fprintf('\ndense qr(A, ''vector''):        %.3f s\n', dt_dense);
fprintf('CQRRPT speedup:               %.1fx\n', dt_dense / dt);
fprintf('leading pivots agree on %d of the first %d columns\n', ...
        numel(intersect(J(1:k), Jd(1:k))), k);

File: python/examples/cqrrpt_demo.py
"""Rank revelation with RandLAPACK CQRRPT.

CQRRPT discovers the numerical rank of a tall matrix rather than taking it as
a parameter. This demo builds a matrix of known rank by repeating columns,
then checks that the detected rank, the pivots, and the residual all agree
with the construction, and compares against SciPy's dense QRCP.

The shape below is chosen where the randomized method actually pays: dense
QRCP wins on small problems (measured 0.6x at 5000x50) and loses as the
matrix grows (7.4x at 20000x200, 2.2x at 50000x1000 on this machine).

Run after installing the package (see bootstrap.sh or README):
    python python/examples/cqrrpt_demo.py
"""
import time

import numpy as np

import randlapack as rl


def main():
    m, r, extra = 20000, 160, 40
    rng = np.random.default_rng(0)
    B = rng.standard_normal((m, r))
    A = np.hstack([B, B[:, :extra]])            # rank r, r+extra columns
    A = np.asfortranarray(A[:, rng.permutation(A.shape[1])])

    t0 = time.perf_counter()
    Q, R, J = rl.cqrrpt(A)
    dt = time.perf_counter() - t0

    k = R.shape[0]
    print(f"A is {m}-by-{A.shape[1]}, constructed rank {r}")
    print(f"CQRRPT detected rank {k} in {dt:.3f} s")
    print(f"||A[:, J] - Q@R|| / ||A||     = "
          f"{np.linalg.norm(A[:, J] - Q @ R) / np.linalg.norm(A):.2e}")
    print(f"||Q.T @ Q - I||               = "
          f"{np.linalg.norm(Q.T @ Q - np.eye(k)):.2e}")

    # J is 0-based here (scipy convention), so it slices A directly.
    print(f"rank(A[:, J[:k]])             = "
          f"{np.linalg.matrix_rank(A[:, J[:k]])}  (want {k})")

    try:
        from scipy.linalg import qr as dense_qr
    except ImportError:
        print("\n(install scipy to see the dense QRCP comparison)")
        return

    t0 = time.perf_counter()
    _, _, Jd = dense_qr(A, mode="economic", pivoting=True)
    dt_dense = time.perf_counter() - t0
    print(f"\nscipy dense QRCP:             {dt_dense:.3f} s")
    print(f"CQRRPT speedup:               {dt_dense / dt:.1f}x")
    print(f"leading pivots agree on {len(set(J[:k]) & set(Jd[:k]))} "
          f"of the first {k} columns")


if __name__ == "__main__":
    main()
